File: apps/backend/app/core/calendar/hijri_date.py
def increment_hijri_day(date):
    if date["day"] < 30:
        return {**date, "day": date["day"] + 1}
    return start_new_month(date)


def decrement_hijri_day(date):
    d, m, y = date["day"], date["month"], date["year"]
    if d > 1:
        return {**date, "day": d - 1}
    pm, py = (m - 1, y) if m > 1 else (12, y - 1)
    return {"year": py, "month": pm, "day": 30}


def start_new_month(date):
    m, y = date["month"], date["year"]
    nm, ny = (m + 1, y) if m < 12 else (1, y + 1)
    return {"year": ny, "month": nm, "day": 1}

File: core/calendar/hijri_date.go
package calendar

import (
	"fmt"
	"time"

	"github.com/ardie069/kalender-hijriyah/core/models"
)

func FormatHijri(h models.HijriDate) string {
	return fmt.Sprintf("%d %s %d", h.Day, h.MonthName, h.Year)
}

// PredictHijriDate: Menghitung transisi tanggal
func PredictHijriDate(t time.Time, isNewMonth bool, currentMonth, currentYear int) models.HijriDate {
	if isNewMonth {
		// Transisi ke bulan baru (Tanggal 1)
		newMonth := (currentMonth % 12) + 1
		newYear := currentYear
		if newMonth == 1 {
			newYear++
		}
		return models.HijriDate{
			Day:       1,
			Month:     newMonth,
			MonthName: MonthNames[newMonth-1],
			Year:      newYear,
		}
	}

	// Gagal ganti bulan -> Bulan sebelumnya digenapkan (Istikmal 30 hari)
	// Kita pake logika: prevMonth = currentMonth
	// Karena kita sedang berada di hari ke-29/30 bulan tersebut.
	return models.HijriDate{
		Day:       30,
		Month:     currentMonth,
		MonthName: MonthNames[currentMonth-1],
		Year:      currentYear,
	}
}
